consensus/src/quorum_store/tests/batch_aggregator_test.rs

```diff
@@ -0,0 +1,301 @@
+    assert_eq!(
+        state_small.append_transactions(all_txns_clone).unwrap_err(),
+        BatchAggregatorError::SizeLimitExceeded
+    );
+    assert_eq!(state_small.num_fragments(), 1);
+
+    assert_eq!(
+        state_small.append_transactions(Vec::new()).unwrap_err(),
+        BatchAggregatorError::SizeLimitExceeded
+    );
+    assert_eq!(state_small.num_fragments(), 2);
+
+    let batch_overflow = state_small.finalize_batch();
+    assert_eq!(
+        batch_overflow.unwrap_err(),
+        BatchAggregatorError::SizeLimitExceeded
+    );
+}
+
+#[test]
+fn test_batch_state_deserialization() {
+    let mut state = IncrementalBatchState::new(1000);
+    let txns = vec![SerializedTransaction::from_bytes(vec![8, 2, 0])];
+
+    assert_eq!(state.num_fragments(), 0);
+    assert_eq!(
+        state.append_transactions(txns).unwrap_err(),
+        BatchAggregatorError::DeserializationError
+    );
+    assert_eq!(state.num_fragments(), 1);
+
+    assert_eq!(
+        state.append_transactions(Vec::new()).unwrap_err(),
+        BatchAggregatorError::DeserializationError
+    );
+    assert_eq!(state.num_fragments(), 2);
+
+    assert_eq!(
+        state
+            .append_transactions(create_vec_serialized_transactions(20))
+            .unwrap_err(),
+        BatchAggregatorError::DeserializationError
+    );
+
+    // Would overflow a non-error state.
+    assert_eq!(state.num_fragments(), 3);
+    let batch = state.finalize_batch();
+    assert_eq!(
+        batch.unwrap_err(),
+        BatchAggregatorError::DeserializationError
+    );
+}
+
+fn check_outdated_fragments(aggregator: &mut BatchAggregator, pairs: Vec<(BatchId, usize)>) {
+    for (i, j) in pairs {
+        assert_eq!(
+            aggregator
+                .append_transactions(i, j, Vec::new())
+                .unwrap_err(),
+            BatchAggregatorError::OutdatedFragment
+        );
+
+        assert_eq!(
+            aggregator.end_batch(i, j, Vec::new()).unwrap_err(),
+            BatchAggregatorError::OutdatedFragment
+        );
+    }
+}
+
+#[test]
+fn test_batch_aggregator_ids() {
+    let mut aggregator = BatchAggregator::new(1000);
+
+    // Forwards batch_id to 3, realizes fragment 0 is skipped.
+    assert_eq!(
+        aggregator
+            .append_transactions(3, 1, Vec::new())
+            .unwrap_err(),
+        BatchAggregatorError::MissedFragment
+    );
+    // Everything <= batch 3 is now outdated.
+    for i in 0..3 {
+        for j in 0..2 {
+            assert_eq!(
+                aggregator
+                    .append_transactions(i, j, Vec::new())
+                    .unwrap_err(),
+                BatchAggregatorError::OutdatedFragment
+            );
+        }
+    }
+
+    assert_ok!(aggregator.append_transactions(4, 0, Vec::new()));
+    assert_ok!(aggregator.append_transactions(4, 1, Vec::new()));
+    assert_ok!(aggregator.append_transactions(4, 2, Vec::new()));
+    check_outdated_fragments(
```

Comment:
Hmm, its interesting that we call (4, 0) and (4, 1) outdated as well, but they are actually in the aggregator.

Reply:
at the time of receiving, those are outdated. in fact, (4, 2) should be as well since we are expecting higher than that, adjusted the test to also check that.